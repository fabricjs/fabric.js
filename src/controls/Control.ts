/* eslint-disable @typescript-eslint/no-unused-vars */
import { halfPI } from '../constants';
import {
  ControlActionHandler,
  TPointerEvent,
  TransformActionHandler,
} from '../EventTypeDefs';
import { Point } from '../Point';
import type { FabricObject } from '../shapes/Object/Object';
import { TDegree, TMat2D } from '../typedefs';
import { cos } from '../util/misc/cos';
import { degreesToRadians } from '../util/misc/radiansDegreesConversion';
import { sin } from '../util/misc/sin';
import {
  ControlRenderingStyleOverride,
  renderCircleControl,
  renderSquareControl,
} from './controls.render';

export class Control {
  /**
   * keep track of control visibility.
   * mainly for backward compatibility.
   * if you do not want to see a control, you can remove it
   * from the control set.
   * @type {Boolean}
   * @default true
   */
  visible = true;

  /**
   * Name of the action that the control will likely execute.
   * This is optional. FabricJS uses to identify what the user is doing for some
   * extra optimizations. If you are writing a custom control and you want to know
   * somewhere else in the code what is going on, you can use this string here.
   * you can also provide a custom getActionName if your control run multiple actions
   * depending on some external state.
   * default to scale since is the most common, used on 4 corners by default
   * @type {String}
   * @default 'scale'
   */
  actionName = 'scale';

  /**
   * Drawing angle of the control.
   * NOT used for now, but name marked as needed for internal logic
   * example: to reuse the same drawing function for different rotated controls
   * @type {Number}
   * @default 0
   */
  angle = 0;

  /**
   * Relative position of the control. X
   * 0,0 is the center of the Object, while -0.5 (left) or 0.5 (right) are the extremities
   * of the bounding box.
   * @type {Number}
   * @default 0
   */
  x = 0;

  /**
   * Relative position of the control. Y
   * 0,0 is the center of the Object, while -0.5 (top) or 0.5 (bottom) are the extremities
   * of the bounding box.
   * @type {Number}
   * @default 0
   */
  y = 0;

  /**
   * Horizontal offset of the control from the defined position. In pixels
   * Positive offset moves the control to the right, negative to the left.
   * It used when you want to have position of control that does not scale with
   * the bounding box. Example: rotation control is placed at x:0, y: 0.5 on
   * the boundind box, with an offset of 30 pixels vertically. Those 30 pixels will
   * stay 30 pixels no matter how the object is big. Another example is having 2
   * controls in the corner, that stay in the same position when the object scale.
   * of the bounding box.
   * @type {Number}
   * @default 0
   */
  offsetX = 0;

  /**
   * Vertical offset of the control from the defined position. In pixels
   * Positive offset moves the control to the bottom, negative to the top.
   * @type {Number}
   * @default 0
   */
  offsetY = 0;

  /**
   * Sets the length of the control. If null, defaults to object's cornerSize.
   * Expects both sizeX and sizeY to be set when set.
   * @type {?Number}
   * @default null
   */
  sizeX: number | null = null;

  /**
   * Sets the height of the control. If null, defaults to object's cornerSize.
   * Expects both sizeX and sizeY to be set when set.
   * @type {?Number}
   * @default null
   */
  sizeY: number | null = null;

  /**
   * Sets the length of the touch area of the control. If null, defaults to object's touchCornerSize.
   * Expects both touchSizeX and touchSizeY to be set when set.
   * @type {?Number}
   * @default null
   */
  touchSizeX: number | null = null;

  /**
   * Sets the height of the touch area of the control. If null, defaults to object's touchCornerSize.
   * Expects both touchSizeX and touchSizeY to be set when set.
   * @type {?Number}
   * @default null
   */
  touchSizeY: number | null = null;

  /**
   * Css cursor style to display when the control is hovered.
   * if the method `cursorStyleHandler` is provided, this property is ignored.
   * @type {String}
   * @default 'crosshair'
   */
  cursorStyle = 'crosshair';

  /**
   * If controls has an offsetY or offsetX, draw a line that connects
   * the control to the bounding box
   * @type {Boolean}
   * @default false
   */
  withConnection = false;

  constructor(options: Partial<Control>) {
    Object.assign(this, options);
  }

  /**
   * The control actionHandler, provide one to handle action ( control being moved )
   * @param {Event} eventData the native mouse event
   * @param {Transform} transformData properties of the current transform
   * @param {Number} x x position of the cursor
   * @param {Number} y y position of the cursor
   * @return {Boolean} true if the action/event modified the object
   */
  declare actionHandler: TransformActionHandler;

  /**
   * The control handler for mouse down, provide one to handle mouse down on control
   * @param {Event} eventData the native mouse event
   * @param {Transform} transformData properties of the current transform
   * @param {Number} x x position of the cursor
   * @param {Number} y y position of the cursor
   * @return {Boolean} true if the action/event modified the object
   */
  declare mouseDownHandler?: ControlActionHandler;

  /**
   * The control mouseUpHandler, provide one to handle an effect on mouse up.
   * @param {Event} eventData the native mouse event
   * @param {Transform} transformData properties of the current transform
   * @param {Number} x x position of the cursor
   * @param {Number} y y position of the cursor
   * @return {Boolean} true if the action/event modified the object
   */
  declare mouseUpHandler?: ControlActionHandler;

  /**
   * Returns control actionHandler
   * @param {Event} eventData the native mouse event
   * @param {FabricObject} fabricObject on which the control is displayed
   * @param {Control} control control for which the action handler is being asked
   * @return {Function} the action handler
   */
  getActionHandler(
    eventData: TPointerEvent,
    fabricObject: FabricObject,
    control: Control
  ): TransformActionHandler | undefined {
    return this.actionHandler;
  }

  /**
   * Returns control mouseDown handler
   * @param {Event} eventData the native mouse event
   * @param {FabricObject} fabricObject on which the control is displayed
   * @param {Control} control control for which the action handler is being asked
   * @return {Function} the action handler
   */
  getMouseDownHandler(
    eventData: TPointerEvent,
    fabricObject: FabricObject,
    control: Control
  ): ControlActionHandler | undefined {
    return this.mouseDownHandler;
  }

  /**
   * Returns control mouseUp handler.
   * During actions the fabricObject or the control can be of different obj
   * @param {Event} eventData the native mouse event
   * @param {FabricObject} fabricObject on which the control is displayed
   * @param {Control} control control for which the action handler is being asked
   * @return {Function} the action handler
   */
  getMouseUpHandler(
    eventData: TPointerEvent,
    fabricObject: FabricObject,
    control: Control
  ): ControlActionHandler | undefined {
    return this.mouseUpHandler;
  }

  /**
   * Returns control cursorStyle for css using cursorStyle. If you need a more elaborate
   * function you can pass one in the constructor
   * the cursorStyle property
   * @param {Event} eventData the native mouse event
   * @param {Control} control the current control ( likely this)
   * @param {FabricObject} object on which the control is displayed
   * @return {String}
   */
  cursorStyleHandler(
    eventData: TPointerEvent,
    control: Control,
    fabricObject: FabricObject
  ) {
    return control.cursorStyle;
  }

  /**
   * Returns the action name. The basic implementation just return the actionName property.
   * @param {Event} eventData the native mouse event
   * @param {Control} control the current control ( likely this)
   * @param {FabricObject} object on which the control is displayed
   * @return {String}
   */
  getActionName(
    eventData: TPointerEvent,
    control: Control,
    fabricObject: FabricObject
  ) {
    return control.actionName;
  }

  /**
   * Returns controls visibility
   * @param {FabricObject} object on which the control is displayed
   * @param {String} controlKey key where the control is memorized on the
   * @return {Boolean}
   */
  getVisibility(fabricObject: FabricObject, controlKey: string) {
    // @ts-expect-error TODO remove directive once fixed
    return fabricObject._controlsVisibility?.[controlKey] ?? this.visible;
  }

  /**
   * Sets controls visibility
   * @param {Boolean} visibility for the object
   * @return {Void}
   */
  setVisibility(visibility: boolean, name: string, fabricObject: FabricObject) {
    this.visible = visibility;
  }

  positionHandler(
    dim: Point,
    finalMatrix: TMat2D,
    fabricObject: FabricObject,
    currentControl: Control
  ) {
    return new Point(
      this.x * dim.x + this.offsetX,
      this.y * dim.y + this.offsetY
    ).transform(finalMatrix);
  }

  /**
   * Returns the coords for this control based on object values.
   * @param {Number} objectAngle angle from the fabric object holding the control
   * @param {Number} objectCornerSize cornerSize from the fabric object holding the control (or touchCornerSize if
   *   isTouch is true)
   * @param {Number} centerX x coordinate where the control center should be
   * @param {Number} centerY y coordinate where the control center should be
   * @param {boolean} isTouch true if touch corner, false if normal corner
   */
  calcCornerCoords(
    objectAngle: TDegree,
    objectCornerSize: number,
    centerX: number,
    centerY: number,
    isTouch: boolean
  ) {
    let cosHalfOffset, sinHalfOffset, cosHalfOffsetComp, sinHalfOffsetComp;
    const xSize = isTouch ? this.touchSizeX : this.sizeX,
      ySize = isTouch ? this.touchSizeY : this.sizeY;
    if (xSize && ySize && xSize !== ySize) {
      // handle rectangular corners
      const controlTriangleAngle = Math.atan2(ySize, xSize);
      const cornerHypotenuse = Math.sqrt(xSize * xSize + ySize * ySize) / 2;
      const newTheta = controlTriangleAngle - degreesToRadians(objectAngle);
      const newThetaComp =
        halfPI - controlTriangleAngle - degreesToRadians(objectAngle);
      cosHalfOffset = cornerHypotenuse * cos(newTheta);
      sinHalfOffset = cornerHypotenuse * sin(newTheta);
      // use complementary angle for two corners
      cosHalfOffsetComp = cornerHypotenuse * cos(newThetaComp);
      sinHalfOffsetComp = cornerHypotenuse * sin(newThetaComp);
    } else {
      // handle square corners
      // use default object corner size unless size is defined
      const cornerSize = xSize && ySize ? xSize : objectCornerSize;
      const cornerHypotenuse = cornerSize * Math.SQRT1_2;
      // complementary angles are equal since they're both 45 degrees
      const newTheta = degreesToRadians(45 - objectAngle);
      cosHalfOffset = cosHalfOffsetComp = cornerHypotenuse * cos(newTheta);
      sinHalfOffset = sinHalfOffsetComp = cornerHypotenuse * sin(newTheta);
    }

    return {
      tl: new Point(centerX - sinHalfOffsetComp, centerY - cosHalfOffsetComp),
      tr: new Point(centerX + cosHalfOffset, centerY - sinHalfOffset),
      bl: new Point(centerX - cosHalfOffset, centerY + sinHalfOffset),
      br: new Point(centerX + sinHalfOffsetComp, centerY + cosHalfOffsetComp),
    };
  }

  /**
   * Render function for the control.
   * When this function runs the context is unscaled. unrotate. Just retina scaled.
   * all the functions will have to translate to the point left,top before starting Drawing
   * if they want to draw a control where the position is detected.
   * left and top are the result of the positionHandler function
   * @param {RenderingContext2D} ctx the context where the control will be drawn
   * @param {Number} left position of the canvas where we are about to render the control.
   * @param {Number} top position of the canvas where we are about to render the control.
   * @param {Object} styleOverride
   * @param {FabricObject} fabricObject the object where the control is about to be rendered
   */
  render(
    ctx: CanvasRenderingContext2D,
    left: number,
    top: number,
    styleOverride: ControlRenderingStyleOverride | undefined,
    fabricObject: FabricObject
  ) {
    styleOverride = styleOverride || {};
    switch (styleOverride.cornerStyle || fabricObject.cornerStyle) {
      case 'circle':
        renderCircleControl.call(
          this,
          ctx,
          left,
          top,
          styleOverride,
          fabricObject
        );
        break;
      default:
        renderSquareControl.call(
          this,
          ctx,
          left,
          top,
          styleOverride,
          fabricObject
        );
    }
  }
}
