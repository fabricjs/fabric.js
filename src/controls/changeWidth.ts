import { TransformActionHandler } from '../EventTypeDefs';
import { getLocalPoint, isTransformCentered } from './util';
import { wrapWithFireEvent } from './wrapWithFireEvent';
import { wrapWithFixedAnchor } from './wrapWithFixedAnchor';

/**
 * Action handler to change object's width
 * Needs to be wrapped with `wrapWithFixedAnchor` to be effective
 * @param {Event} eventData javascript event that is doing the transform
 * @param {Object} transform javascript object containing a series of information around the current transform
 * @param {number} x current mouse x position, canvas normalized
 * @param {number} y current mouse y position, canvas normalized
 * @return {Boolean} true if some change happened
 */
export const changeObjectWidth: TransformActionHandler = (
  eventData,
  transform,
  x,
  y
) => {
  const localPoint = getLocalPoint(
    transform,
    transform.originX,
    transform.originY,
    x,
    y
  );
  //  make sure the control changes width ONLY from it's side of target
  if (
    transform.originX === 'center' ||
    (transform.originX === 'right' && localPoint.x < 0) ||
    (transform.originX === 'left' && localPoint.x > 0)
  ) {
    const { target } = transform,
      strokePadding =
        target.strokeWidth / (target.strokeUniform ? target.scaleX : 1),
      multiplier = isTransformCentered(transform) ? 2 : 1,
      oldWidth = target.width,
      newWidth = Math.ceil(
        Math.abs((localPoint.x * multiplier) / target.scaleX) - strokePadding
      );
    target.set('width', Math.max(newWidth, 0));
    //  check against actual target width in case `newWidth` was rejected
    return oldWidth !== target.width;
  }
  return false;
};

export const changeWidth = wrapWithFireEvent(
  'resizing',
  wrapWithFixedAnchor(changeObjectWidth)
);
