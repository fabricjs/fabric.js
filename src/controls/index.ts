export { changeWidth } from './changeWidth';
export { renderCircleControl, renderSquareControl } from './controlRendering';
export * from './commonControls';
export { dragHandler } from './drag';
export * from './polyControl';
export { rotationStyleHandler, rotationWithSnapping } from './rotate';
export {
  scaleCursorStyleHandler,
  scalingEqually,
  scalingX,
  scalingY,
} from './scale';
export {
  scaleOrSkewActionName,
  scaleSkewCursorStyleHandler,
  scalingXOrSkewingY,
  scalingYOrSkewingX,
} from './scaleSkew';
export { skewCursorStyleHandler, skewHandlerX, skewHandlerY } from './skew';
export { getLocalPoint } from './util';
export { wrapWithFireEvent } from './wrapWithFireEvent';
export { wrapWithFixedAnchor } from './wrapWithFixedAnchor';
