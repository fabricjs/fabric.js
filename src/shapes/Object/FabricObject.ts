import { ObjectEvents } from '../../EventTypeDefs';
import { FabricObjectSVGExportMixin } from './FabricObjectSVGExportMixin';
import { InteractiveFabricObject } from './InteractiveObject';
import { applyMixins } from '../../util/applyMixins';
import { StatefulMixin } from '../../mixins/stateful.mixin';

// TODO somehow we have to make a tree-shakeable import

// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unused-vars
export interface FabricObject<EventSpec extends ObjectEvents = ObjectEvents>
  extends FabricObjectSVGExportMixin {}

export class FabricObject<
  EventSpec extends ObjectEvents = ObjectEvents
> extends InteractiveFabricObject<EventSpec> {}

applyMixins(FabricObject, [FabricObjectSVGExportMixin, StatefulMixin]);

export { cacheProperties, stateProperties } from './defaultValues';
