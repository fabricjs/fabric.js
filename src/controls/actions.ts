import { fabric } from '../../HEADER';
import { changeWidth } from './changeWidth';
import { renderCircleControl, renderSquareControl } from './controls.render';
import { dragHandler } from './drag';
import { rotationStyleHandler, rotationWithSnapping } from './rotate';
import {
  scaleCursorStyleHandler,
  scalingEqually,
  scalingX,
  scalingY,
} from './scale';
import {
  scaleOrSkewActionName,
  scaleSkewCursorStyleHandler,
  scalingXOrSkewingY,
  scalingYOrSkewingX,
} from './scaleSkew';
import { skewCursorStyleHandler, skewHandlerX, skewHandlerY } from './skew';
import { getLocalPoint, getActionFromCorner } from './util';
import { wrapWithFireEvent } from './wrapWithFireEvent';
import { wrapWithFixedAnchor } from './wrapWithFixedAnchor';

export {
  scaleCursorStyleHandler,
  skewCursorStyleHandler,
  scaleSkewCursorStyleHandler,
  rotationWithSnapping,
  scalingEqually,
  scalingX,
  scalingY,
  scalingYOrSkewingX,
  scalingXOrSkewingY,
  changeWidth,
  skewHandlerX,
  skewHandlerY,
  dragHandler,
  scaleOrSkewActionName,
  rotationStyleHandler,
  wrapWithFixedAnchor,
  wrapWithFireEvent,
  getLocalPoint,
  getActionFromCorner,
  renderCircleControl,
  renderSquareControl,
};

/**
 * @todo remove as unused
 */
fabric.controlsUtils = {
  scaleCursorStyleHandler,
  skewCursorStyleHandler,
  scaleSkewCursorStyleHandler,
  rotationWithSnapping,
  scalingEqually,
  scalingX,
  scalingY,
  scalingYOrSkewingX,
  scalingXOrSkewingY,
  changeWidth,
  skewHandlerX,
  skewHandlerY,
  dragHandler,
  scaleOrSkewActionName,
  rotationStyleHandler,
  wrapWithFixedAnchor,
  wrapWithFireEvent,
  getLocalPoint,
  renderCircleControl,
  renderSquareControl,
};
