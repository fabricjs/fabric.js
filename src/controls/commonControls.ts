import { changeWidth } from './changeWidth';
import { Control } from './Control';
import { rotationStyleHandler, rotationWithSnapping } from './rotate';
import { scaleCursorStyleHandler, scalingEqually } from './scale';
import {
  scaleOrSkewActionName,
  scaleSkewCursorStyleHandler,
  scalingXOrSkewingY,
  scalingYOrSkewingX,
} from './scaleSkew';

// use this function if you want to generate new controls for every instance
export const createObjectDefaultControls = () => ({
  ml: new Control({
    x: -0.5,
    y: 0,
    cursorStyleHandler: scaleSkewCursorStyleHandler,
    actionHandler: scalingXOrSkewingY,
    getActionName: scaleOrSkewActionName,
  }),

  mr: new Control({
    x: 0.5,
    y: 0,
    cursorStyleHandler: scaleSkewCursorStyleHandler,
    actionHandler: scalingXOrSkewingY,
    getActionName: scaleOrSkewActionName,
  }),

  mb: new Control({
    x: 0,
    y: 0.5,
    cursorStyleHandler: scaleSkewCursorStyleHandler,
    actionHandler: scalingYOrSkewingX,
    getActionName: scaleOrSkewActionName,
  }),

  mt: new Control({
    x: 0,
    y: -0.5,
    cursorStyleHandler: scaleSkewCursorStyleHandler,
    actionHandler: scalingYOrSkewingX,
    getActionName: scaleOrSkewActionName,
  }),

  tl: new Control({
    x: -0.5,
    y: -0.5,
    cursorStyleHandler: scaleCursorStyleHandler,
    actionHandler: scalingEqually,
  }),

  tr: new Control({
    x: 0.5,
    y: -0.5,
    cursorStyleHandler: scaleCursorStyleHandler,
    actionHandler: scalingEqually,
  }),

  bl: new Control({
    x: -0.5,
    y: 0.5,
    cursorStyleHandler: scaleCursorStyleHandler,
    actionHandler: scalingEqually,
  }),

  br: new Control({
    x: 0.5,
    y: 0.5,
    cursorStyleHandler: scaleCursorStyleHandler,
    actionHandler: scalingEqually,
  }),

  mtr: new Control({
    x: 0,
    y: -0.5,
    actionHandler: rotationWithSnapping,
    cursorStyleHandler: rotationStyleHandler,
    offsetY: -40,
    withConnection: true,
    actionName: 'rotate',
  }),
});

export const createResizeControls = () => ({
  mr: new Control({
    x: 0.5,
    y: 0,
    actionHandler: changeWidth,
    cursorStyleHandler: scaleSkewCursorStyleHandler,
    actionName: 'resizing',
  }),
  ml: new Control({
    x: -0.5,
    y: 0,
    actionHandler: changeWidth,
    cursorStyleHandler: scaleSkewCursorStyleHandler,
    actionName: 'resizing',
  }),
});

export const createTextboxDefaultControls = () => ({
  ...createObjectDefaultControls(),
  ...createResizeControls(),
});
