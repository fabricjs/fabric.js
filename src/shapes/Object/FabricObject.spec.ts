import { roundSnapshotOptions } from '../../../jest.extend';
import { createObjectDefaultControls } from '../../controls';
import { Group } from '../Group';
import { FabricObject } from './FabricObject';

describe('Object', () => {
  test('setCoords', () => {
    const controls = createObjectDefaultControls();
    const object = new FabricObject({
      left: 25,
      top: 60,
      width: 75,
      height: 100,
      controls,
      scaleY: 2,
      fill: 'blue',
    });
    const group = new Group([object], {
      angle: 30,
      scaleX: 2,
      interactive: true,
      subTargetCheck: true,
    });
    const spies = Object.values(controls).map((control) =>
      jest.spyOn(control, 'calcCornerCoords')
    );
    group.setCoords();
    expect(
      Object.keys(controls).reduce((acc, key, index) => {
        const { calls, results } = spies[index].mock;
        return Object.assign(acc, { [key]: { calls, results } });
      }, {})
    ).toMatchSnapshot(roundSnapshotOptions);
  });
});
