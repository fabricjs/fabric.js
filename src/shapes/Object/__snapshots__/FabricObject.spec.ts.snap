// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object setCoords 1`] = `
{
  "bl": {
    "calls": [
      [
        30,
        13,
        -51,
        235,
        false,
      ],
      [
        30,
        24,
        -51,
        235,
        true,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": -60,
            "y": 237,
          },
          "br": Point {
            "x": -49,
            "y": 244,
          },
          "tl": Point {
            "x": -53,
            "y": 226,
          },
          "tr": Point {
            "x": -42,
            "y": 233,
          },
        },
      },
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": -67,
            "y": 239,
          },
          "br": Point {
            "x": -47,
            "y": 251,
          },
          "tl": Point {
            "x": -55,
            "y": 219,
          },
          "tr": Point {
            "x": -35,
            "y": 231,
          },
        },
      },
    ],
  },
  "br": {
    "calls": [
      [
        30,
        13,
        81,
        311,
        false,
      ],
      [
        30,
        24,
        81,
        311,
        true,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 72,
            "y": 313,
          },
          "br": Point {
            "x": 83,
            "y": 320,
          },
          "tl": Point {
            "x": 78,
            "y": 302,
          },
          "tr": Point {
            "x": 90,
            "y": 309,
          },
        },
      },
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 64,
            "y": 315,
          },
          "br": Point {
            "x": 85,
            "y": 327,
          },
          "tl": Point {
            "x": 76,
            "y": 295,
          },
          "tr": Point {
            "x": 97,
            "y": 307,
          },
        },
      },
    ],
  },
  "mb": {
    "calls": [
      [
        30,
        13,
        15,
        273,
        false,
      ],
      [
        30,
        24,
        15,
        273,
        true,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 6,
            "y": 275,
          },
          "br": Point {
            "x": 17,
            "y": 282,
          },
          "tl": Point {
            "x": 12,
            "y": 264,
          },
          "tr": Point {
            "x": 24,
            "y": 271,
          },
        },
      },
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": -2,
            "y": 277,
          },
          "br": Point {
            "x": 19,
            "y": 289,
          },
          "tl": Point {
            "x": 10,
            "y": 257,
          },
          "tr": Point {
            "x": 31,
            "y": 269,
          },
        },
      },
    ],
  },
  "ml": {
    "calls": [
      [
        30,
        13,
        -0,
        147,
        false,
      ],
      [
        30,
        24,
        -0,
        147,
        true,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": -9,
            "y": 150,
          },
          "br": Point {
            "x": 2,
            "y": 156,
          },
          "tl": Point {
            "x": -3,
            "y": 139,
          },
          "tr": Point {
            "x": 8,
            "y": 145,
          },
        },
      },
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": -17,
            "y": 152,
          },
          "br": Point {
            "x": 4,
            "y": 164,
          },
          "tl": Point {
            "x": -5,
            "y": 131,
          },
          "tr": Point {
            "x": 16,
            "y": 143,
          },
        },
      },
    ],
  },
  "mr": {
    "calls": [
      [
        30,
        13,
        131,
        223,
        false,
      ],
      [
        30,
        24,
        131,
        223,
        true,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 122,
            "y": 226,
          },
          "br": Point {
            "x": 134,
            "y": 232,
          },
          "tl": Point {
            "x": 129,
            "y": 215,
          },
          "tr": Point {
            "x": 140,
            "y": 221,
          },
        },
      },
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 115,
            "y": 228,
          },
          "br": Point {
            "x": 136,
            "y": 240,
          },
          "tl": Point {
            "x": 127,
            "y": 207,
          },
          "tr": Point {
            "x": 148,
            "y": 219,
          },
        },
      },
    ],
  },
  "mt": {
    "calls": [
      [
        30,
        13,
        116,
        98,
        false,
      ],
      [
        30,
        24,
        116,
        98,
        true,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 107,
            "y": 100,
          },
          "br": Point {
            "x": 118,
            "y": 107,
          },
          "tl": Point {
            "x": 113,
            "y": 89,
          },
          "tr": Point {
            "x": 125,
            "y": 96,
          },
        },
      },
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 99,
            "y": 102,
          },
          "br": Point {
            "x": 120,
            "y": 114,
          },
          "tl": Point {
            "x": 111,
            "y": 82,
          },
          "tr": Point {
            "x": 132,
            "y": 94,
          },
        },
      },
    ],
  },
  "mtr": {
    "calls": [
      [
        30,
        13,
        136,
        63,
        false,
      ],
      [
        30,
        24,
        136,
        63,
        true,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 127,
            "y": 66,
          },
          "br": Point {
            "x": 138,
            "y": 72,
          },
          "tl": Point {
            "x": 133,
            "y": 54,
          },
          "tr": Point {
            "x": 145,
            "y": 61,
          },
        },
      },
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 119,
            "y": 68,
          },
          "br": Point {
            "x": 140,
            "y": 80,
          },
          "tl": Point {
            "x": 131,
            "y": 47,
          },
          "tr": Point {
            "x": 152,
            "y": 59,
          },
        },
      },
    ],
  },
  "tl": {
    "calls": [
      [
        30,
        13,
        50,
        60,
        false,
      ],
      [
        30,
        24,
        50,
        60,
        true,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 41,
            "y": 62,
          },
          "br": Point {
            "x": 52,
            "y": 69,
          },
          "tl": Point {
            "x": 48,
            "y": 51,
          },
          "tr": Point {
            "x": 59,
            "y": 58,
          },
        },
      },
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 34,
            "y": 64,
          },
          "br": Point {
            "x": 54,
            "y": 76,
          },
          "tl": Point {
            "x": 46,
            "y": 44,
          },
          "tr": Point {
            "x": 66,
            "y": 56,
          },
        },
      },
    ],
  },
  "tr": {
    "calls": [
      [
        30,
        13,
        182,
        136,
        false,
      ],
      [
        30,
        24,
        182,
        136,
        true,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 173,
            "y": 138,
          },
          "br": Point {
            "x": 184,
            "y": 145,
          },
          "tl": Point {
            "x": 179,
            "y": 127,
          },
          "tr": Point {
            "x": 191,
            "y": 134,
          },
        },
      },
      {
        "type": "return",
        "value": {
          "bl": Point {
            "x": 165,
            "y": 140,
          },
          "br": Point {
            "x": 186,
            "y": 152,
          },
          "tl": Point {
            "x": 177,
            "y": 120,
          },
          "tr": Point {
            "x": 198,
            "y": 132,
          },
        },
      },
    ],
  },
}
`;
